proposals/distributed-async-epoch-reward.md

```diff
@@ -0,0 +1,294 @@
+boundary by dividing the process into two distinct phases:
+
+   1. rewards calculation phase - during which the epoch rewards for all
+      activate stake accounts are computed and distribution chunks are scheduled.
+
+   1. rewards distribution phase - during which the calculated epoch rewards
+      for the active stake accounts are distributed.
+
+### Rewards Calculation
+
+Reward calculation phase computes all the rewards that need to be distributed
+for the active stake accounts, and partitions the reward into a number of
+chunks. Each of the chunks will be distributed in a separate block in the
+`reward distribution phase`.
+
+To help track and verify the reward distribution in the next phase, two new
+sysvar accounts are introduced: (a) `EpochRewardHistory` sysvar account and (b)
+`RewardReserve` sysvar accounts. These sysvar accounts are updated at the
+boundary blocks (i.e. first and last blocks) in the reward calculation phase.
+More details of these two sysvar accounts are described in the following
+sections.
+
+The reward computation phase lasts *exactly* for `N` block heights since
+the beginning of the epoch. When reaching block height `N` after the start
+block of the `reward calculation phase`, the bank will mark it the end the
+`reward calculation phase`.
+
+On Solana Mainnet Beta, the 90% cut-off of the epoch reward computation time
+for all the nodes is around 40 seconds for the ~550K active stake accounts.
+This is approximately 1000 blocks at the cluster's average block rate (40ms
+per block). Therefore, a conservative value of 1,000 is chosen for `N`.
+
+At the end of the `reward calculation phase`, aka. `Epoch_start + N` block
+height, the reward computation process is "barrier synchronized" with the bank
+replay process. The bank replay process won't replaying any blocks after
+`Epoch_start+N` unless the reward computation has completed. If the reward
+calculation is slow and hasn't complete the computation at block-height
+`Epoch_start + N`, the bank replay process will block and wait until the reward
+computation is finished, before processing any blocks with height greater than
+`N`.
+
+After the full rewards result is calculated, the result is partitioned into
+distribution chunks, which will be distributed during the `reward distribution`
+phase.
+
+To ensure that each block distributes a subset of the rewards in a
+deterministic manner for the current epoch, while also randomizing the
+distribution across different epochs, the partitioning of all rewards will be
+done as follows. First, a new key is generated for each reward by hashing the
+account's public key with the epoch number (i.e. sha256 hash from
+solana_sdk::program::hashv). Then, the rewards are sorted by the new key and
+divided into `M` bins. This process will ensure that the reward distribution is
+deterministic for the current epoch, while also introducing a degree of
+randomness between epochs.
```

Comment:
however this will make an O(1) distribution block lookup for any individual account impossible. something the proposal similarly precludes as well... :thinking: